src/digests.rs

```diff
@@ -0,0 +1,213 @@
+        self.0.as_ref()
+    }
+}
+
+impl Multihash {
+    /// Creates a new `Multihash` from a `Vec<u8>`, consuming it.
+    /// If the input data is not a valid  multihash an error is returned.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use multihash::{Sha2256, MultihashDigest, Multihash, Code};
+    ///
+    /// let mh = Sha2256::digest(b"hello world");
+    ///
+    /// // valid multihash
+    /// let mh2 = Multihash::from_vec(mh.into_vec()).unwrap();
+    ///
+    /// // invalid multihash
+    /// assert!(Multihash::from_vec(vec![1,2,3]).is_err());
+    /// ```
+    pub fn from_vec(raw: Vec<u8>) -> Result<Self, Error> {
+        // validate code
+        let (raw_code, code_size) = u32::decode_var(&raw[..]);
+        Code::from_u32(raw_code).ok_or_else(|| Error::Invalid)?;
+
+        // validate size
+        let (size, size_size) = u64::decode_var(&raw[code_size..]);
+        if size != (raw.len() - code_size - size_size) as u64 {
+            return Err(Error::BadInputLength);
+        }
+
+        Ok(Multihash(raw.into_boxed_slice()))
+    }
+
+    /// Creates a new `Multihash` from a slice.
+    /// If the input data is not a valid  multihash an error is returned.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use multihash::{Sha2256, MultihashDigest, Multihash, Code};
+    ///
+    /// let mh = Sha2256::digest(b"hello world");
+    ///
+    /// // valid multihash
+    /// let mh2 = Multihash::from_slice(mh.as_ref()).unwrap();
+    ///
+    /// // invalid multihash
+    /// assert!(Multihash::from_slice(&vec![1,2,3]).is_err());
+    /// ```
+    pub fn from_slice(raw: &[u8]) -> Result<Self, Error> {
+        Multihash::from_vec(raw.into())
+    }
+
+    /// Turns the `Multihash` into a `Vec<u8>`, consuming it.
+    pub fn into_vec(self) -> Vec<u8> {
+        self.0.into_vec()
+    }
+
+    /// Creates a new `Vec<u8>`.
+    pub fn to_vec(&self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+
+    /// Returns the `Code` of this multihash.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use multihash::{Sha2256, MultihashDigest, Code};
+    ///
+    /// let mh = Sha2256::digest(b"hello world");
+    /// assert_eq!(mh.code(), Code::Sha2256);
+    /// ```
+    pub fn code(&self) -> Code {
+        let (raw_code, _) = u32::decode_var(&self.0);
+        Code::from_u32(raw_code).unwrap()
+    }
+
+    /// Returns the algorithm used in this multihash as a string.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use multihash::{Sha2256, MultihashDigest};
+    ///
+    /// let mh = Sha2256::digest(b"hello world");
+    /// assert_eq!(mh.algorithm(), "Sha2256");
+    /// ```
+    pub fn algorithm(&self) -> &'static str {
+        self.code().into()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, MultihashDigest)]
+#[repr(u32)]
+pub enum Code {
+    #[Size = "20"]
+    #[Digest = "sha1::Sha1"]
+    Sha1 = 0x11,
+
+    #[Size = "32"]
+    #[Digest = "sha2::Sha256"]
+    Sha2256 = 0x12,
```

Comment:
Does it really have to be like that?

Reply:
which part?

Reply:
I mean the `Sha2256`. I understand need for no `_` but something like `Sha256V2`,  `Sha256V3`, `Sha512V2`. Anything to make it valid syntactically.

Reply:
@dignifiedquire however you feel, anyway, just an observation.

Reply:
Can we get it to be `SHA2_256` etc.?